src/ray/gcs/gcs_client/gcs_client.h

```diff
@@ -241,6 +264,26 @@ class RAY_EXPORT GcsClient : public std::enable_shared_from_this<GcsClient> {
 
   /// Retry interval to reconnect to a GCS server.
   const int64_t kGCSReconnectionRetryIntervalMs = 1000;
+
+  // Protect the `reconnect_in_progress_` variable.
+  mutable absl::Mutex reconnect_flag_mutex_;
+
+  // A flag to indicate whether reconnecting to gcs is in progress or not.
+  // If false, reconnecting will be triggerd. If true, the current reconnect
+  // request should be ignored.
+  std::atomic<bool> reconnect_in_progress_ GUARDED_BY(reconnect_flag_mutex_) = false;
+
+  // Protect the `callbacks_` list.
+  mutable absl::Mutex reconnect_callbacks_mutex_;
```

Comment:
I believe this callbacks are only added / flushed in the io service of periodical runner. Is this wrong?